### PR TITLE
Add completions to path

### PR DESCRIPTION
**Pull request by nhynes**

### CODE REVIEW

File: installer.py [hunk 1]
```diff
@@ -34,16 +34,18 @@ def main():
 
     args = _parse_args()
 
+    shell = run('ps -p $$ -oargs=', capture=True, shell=True)
```
Comment (ssavvides) [inline installer.py]: This does not seem to work. Tested under OSx and Linux. The output is, e.g., "ps -p 51898 -oargs=".

**Comment by nhynes** [comment]
> there are other ways to decide what shell to modify

elaborate?

**Comment by pro-wh** [comment]
> > there are other ways to decide what shell to modify
> 
> elaborate?

I'm thinking of in Unix systems, each user is associated with a specific shell, and there's this whole `chsh` system for managing it.

Since we're talking about this, I also want to ask about "not pollute other shells' profiles." Why are we so careful to change only one shell's rc?

**Comment by nhynes** [comment]
> I'm thinking of in Unix systems, each user is associated with a specific shell, and there's this whole chsh system for managing it.

Counterpoint: when on a restricted environment (e.g., academic machines), you might configure your terminal emulator to use a different login shell automatically. The `ps -p $$` approach is more likely to do something reasonable.

> Why are we so careful to change only one shell's rc?

Rustup just adds to all profiles. I don't use bash and I don't want a `.profile` or `.bashrc` in my home directory. It's just sloppy.

File: installer.py [hunk 2]
```diff
@@ -209,29 +209,35 @@ def is_oasis(path):
...
+def modify_shell_profile(plat, args, shell):
     """Adds the Oasis tools to the user's PATH via a profile file.
        Assumes that the current shell is the user's preferred shell
        so to not pollute other shells' profiles."""
-    shell = run('ps -p $$ -oargs=', capture=True, shell=True)
     if 'zsh' in shell:
         rcfile = osp.join(os.environ.get('ZDOTDIR', '~'), '.zprofile')
     elif 'bash' in shell:
-        rcfile = '~/.bash_profile'
+        rcfile = '~/.bashrc'
```
Comment (pro-wh) [inline installer.py]: alright, it seems pretty normal for .profile to run .bashrc https://sources.debian.org/src/bash/5.0-4/debian/skel.profile/#L15

but are there users now that ran the earlier version of this and it created a .bash_profile for them and now bash won't run .profile for their login sessions and won't run .bashrc?

File: installer.py [hunk 3]
```diff
@@ -209,29 +209,35 @@ def is_oasis(path):
...
     """Returns the env exports required to run the Oasis toolchain."""
     path_export = 'export PATH=%s/bin:${CARGO_HOME:-~/.cargo}/bin:$PATH' % args.prefix
     exports = [path_export]
     ld_path_key = '%s_LIBRARY_PATH' % ('DYLD' if plat == PLAT_DARWIN else 'LD')
     if RUST_SYSROOT_PREFIX not in os.environ.get(ld_path_key, ''):
         exports.append('export {0}=$(rustc --print sysroot)/lib:{0}'.format(ld_path_key))
+
+    data_dir = osp.join(args.prefix, 'share', 'oasis')
+    if 'zsh' in shell:
+        exports.append('fpath=("%s" $fpath)' % data_dir)
```
Comment (pro-wh) [inline installer.py]: is there anything other than completions.sh that becomes autoloaded this way?

File: installer.py (hunk 2 above)
Comment (pro-wh) [reply installer.py]: oh but also Debian has this at the beginning of their skel .bashrc:
```
# If not running interactively, don't do anything
case $- in
    *i*) ;;
      *) return;;
esac
```

so it won't run for noninteractive sessions:

```
ssh devmachine 'type oasis'
```

Comment (pro-wh) [reply installer.py]: whatever, NVM puts its stuff in .bashrc

File: installer.py [hunk 4]
```diff
@@ -209,29 +209,35 @@ def is_oasis(path):
...
+    else:
```
Comment (pro-wh) [inline installer.py]: can we put this under `if 'bash' in shell:` so that we don't go adding completions to .profile

File: installer.py (hunk 3 above)
Comment (nhynes) [reply installer.py]: No

File: installer.py (hunk 4 above)
Comment (nhynes) [reply installer.py]: what about dash or sh or ksh

File: installer.py (hunk 1 above)
Comment (pro-wh) [reply installer.py]: resolving because we don't do this now

**Comment by nhynes** [comment]
> requesting changes: need a copy of completions.sh and directory listing of prefix/share/oasis to review


I would love to, but it depends on https://github.com/oasislabs/oasis-cli/pull/105

**Comment by nhynes** [comment]
@pro-wh 

```
$ ls ~/.local/share/oasis
_oasis  completions.sh  installed_release  metrics.jsonl

$ cat ~/.local/share/oasis/completions.sh
_oasis() {
    local i cur prev opts cmds
    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    cmd=""
    opts=""

    for i in ${COMP_WORDS[@]}
    do
        case "${i}" in
            oasis)
                cmd="oasis"
                ;;

            build)
                cmd+="__build"
                ;;
            chain)
                cmd+="__chain"
                ;;
            clean)
                cmd+="__clean"
                ;;
            config)
                cmd+="__config"
                ;;
            deploy)
                cmd+="__deploy"
                ;;
            gen_completions)
                cmd+="__gen_completions"
                ;;
            help)
                cmd+="__help"
                ;;
            ifextract)
                cmd+="__ifextract"
                ;;
            init)
                cmd+="__init"
                ;;
            set-toolchain)
                cmd+="__set__toolchain"
                ;;
            test)
                cmd+="__test"
                ;;
            upload_metrics)
                cmd+="__upload_metrics"
                ;;
            *)
                ;;
        esac
    done

    case "${cmd}" in
        oasis)
            opts=" -h -V  --help --version   init build test clean chain ifextract deploy config upload_metrics gen_completions set-toolchain help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;

        oasis__build)
            opts=" -v -q -h -V -s  --debug --verbose --quiet --wasi --help --version --stack  <SERVICE>... <builder_args>... "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                --stack)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -s)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        oasis__chain)
            opts=" -h -V  --help --version  <chain_args>... "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        oasis__clean)
            opts=" -h -V  --help --version  "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        oasis__config)
            opts=" -h -V  --help --version  <KEY> <VALUE> "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        oasis__deploy)
            opts=" -v -q -h -V -p  --verbose --quiet --help --version --profile  <deployer_args>... "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                --profile)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -p)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        oasis__gen_completions)
            opts=" -h -V  --help --version  "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        oasis__help)
            opts=" -h -V  --help --version  "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        oasis__ifextract)
            opts=" -h -V -o  --help --version --out  <SERVICE_URL> "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                --out)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -o)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        oasis__init)
            opts=" -q -h -V  --quiet --rust --help --version  <NAME> "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        oasis__set__toolchain)
            opts=" -h -V  --help --version  <VERSION> "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        oasis__test)
            opts=" -v -q -h -V -p  --verbose --quiet --debug --help --version --profile  <SERVICE>... <tester_args>... "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                --profile)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -p)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        oasis__upload_metrics)
            opts=" -h -V  --help --version  "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in

                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
    esac
}

complete -F _oasis -o bashdefault -o default oasis
```
but I wouldn't worry about the contents of completions.sh because it's auto-generated by clap.

File: installer.py [hunk 5]
```diff
@@ -209,31 +209,37 @@ def is_oasis(path):
...
     elif 'bash' in shell:
-        rcfile = '~/.bash_profile'
+        rcfile = '~/.bashrc'
     else:
         rcfile = '~/.profile'
-
-    required_exports = get_required_exports(plat, args)
     rc_file = osp.expanduser(rcfile)
 
+    required_exports = get_shell_additions(plat, args, shell)
```
Comment (pro-wh) [inline installer.py]: should we change this variable name too now that we've changed the function name?

**Pull request merged**